### PR TITLE
OCPBUGS-15934: use *resource.Quantity to not automatically set 0 

### PR DESCRIPTION
**Pull request by QiWang19**
close: https://issues.redhat.com/browse/OCPBUGS-15934
**- What I did**
Change the type of `OverlaySize` and `LogSizeMax` resource.Quantity to pointer *resource.Quantity. 
the struct type will be set as 0 automatically when retrieving the ctrcfg object.

I am uncertain if we can make this update to current ContainerruntimeConfig API, I propose this change because the ContainerruntimeConfig API is currently MCO internal.
**- How to verify it**
```
# apply the containerruntimeconfig CR
apiVersion: machineconfiguration.openshift.io/v1
kind: ContainerRuntimeConfig
metadata:
 name: pidlimit
spec:
 machineConfigPoolSelector:
   matchLabels:
     pools.operator.machineconfiguration.openshift.io/worker: '' 
 containerRuntimeConfig:
   pidsLimit: 4096 
   logLevel: debug

# current result:
$ oc get containerruntimeconfig  pidlimit -o json | jq '.spec.containerRuntimeConfig'
{
  "logLevel": "debug",
  "logSizeMax": "0",
  "overlaySize": "0",
  "pidsLimit": 4096
}

# after this patch, it will be 
$ oc get containerruntimeconfig  pidlimit -o json | jq '.spec.containerRuntimeConfig'
{
  "logLevel": "debug",
  "pidsLimit": 4096
}
```
**- Description for the changelog**

### CODE REVIEW

**Comment by QiWang19** [comment]
@mtrmac could you take a look?

File: pkg/controller/container-runtime-config/container_runtime_config_bootstrap.go
Comment (mtrmac) [inline pkg/controller/container-runtime-config/container_runtime_config_bootstrap.go:43]: Applies to all changes:

Doesn’t this change the semantics of previously-created CRs? If I understand the situation correctly, it’s possible that users create the CRs with these fields missing, and something (`patchContainerRuntimeConfigs`??) updates them and adds `"0"` field values.

In the old version, those 0 values were treated as missing, i.e. the system did exactly what the users wanted, just in a confusing way; with this PR, wouldn’t the `0` values actually start being applied?

I suppose one way to tell would be to add (uint? e2e?) tests with CRs of that kind, and ensure that both empty values and `"0"` values are treated as missing. (Or maybe those tests already exist? I didn’t find them though a quick grep but I didn’t spend much time looking.)

File: pkg/controller/container-runtime-config/container_runtime_config_controller_test.go
Comment (mtrmac) [inline pkg/controller/container-runtime-config/container_runtime_config_controller_test.go:1399]: Testing `LogSizeMax` might be worthwhile here.

**Review by mtrmac — COMMENTED** [review]
I’m afraid I’m quite unfamiliar with this code.

Looking at the history, https://github.com/openshift/machine-config-operator/pull/2494 supposedly fixed the same bug just by adding the `omitempty` annotations. Is it known what has changed that this is no longer sufficient? (I can’t, from a quick check, notice anything in the history of `resource.Quantity` changing the semantics.)

Or did that never work?

File: pkg/controller/container-runtime-config/container_runtime_config_bootstrap.go
Comment (QiWang19) [reply pkg/controller/container-runtime-config/container_runtime_config_bootstrap.go:43]: 0 values will not be applied in this PR. The drop-in files will not be created if `0` values. https://github.com/openshift/machine-config-operator/pull/3815/files#diff-b2ef9db0d76c6145ec1bb5310134037bae64572ab47464d3223745c2a410d3e3R301

Comment (mtrmac) [reply pkg/controller/container-runtime-config/container_runtime_config_bootstrap.go:43]: Oops, my mistake, I didn’t notice that logic.

`OverlaySize` will, AFAICS, still cause `mergeConfigChanges` to be called, though with no edits. I guess that doesn’t make a difference.

---

Aesthetically I’d prefer for the code to consistently use a single condition, so that we just have “set to a relevant value / not set to a relevant value”, not “unset / set to zero / set to non-zero” to track.

But that’s weak a code style preference, to be decided by MCO maintainers.

**Comment by QiWang19** [comment]
The automatically zero values are set only on some builds. I feel it caused  by different json packages somewhere.

4.14.0-ec.4
```
$ oc version
Client Version: v4.2.0-alpha.0-1974-g13225e0
Kustomize Version: v5.0.1
Server Version: 4.14.0-ec.4
Kubernetes Version: v1.27.3+4aaeaec
$ oc get containerruntimeconfig.machineconfiguration.openshift.io/pidlimit -o json | jq '.spec.containerRuntimeConfig'
{
  "logLevel": "debug",
  "logSizeMax": "0",
  "overlaySize": "0",
  "pidsLimit": 4096
}
```
 

 

 4.14.0-0.ci.:
```
$ oc version
Client Version: v4.2.0-alpha.0-1974-g13225e0
Kustomize Version: v5.0.1
Server Version: 4.14.0-0.ci.test-2023-08-02-171351-ci-ln-q32sdzk-latest
Kubernetes Version: v1.27.1-3201+ba1825544533d2-dirty
$ oc get containerruntimeconfig.machineconfiguration.openshift.io/pidlimit -o json | jq '.spec.containerRuntimeConfig'
{
  "logLevel": "debug",
  "pidsLimit": 4096
}
```

**Comment by mtrmac** [comment]
That’s basically the thing that worries me — if we don’t understand how the bug arose, are we certain that we fixed it?

E.g. what if the `ContainerRuntimeConfig` object in question is created by something not in the MCO codebase? It’s quite possible that we need the MCO change anyway (if the old bug fix was just not working), but to actually fix the reporter’s problem we might need to find and fix that other non-MCO writer.

**Comment by QiWang19** [comment]
> That’s basically the thing that worries me — if we don’t understand how the bug arose, are we certain that we fixed it?
> 
> E.g. what if the `ContainerRuntimeConfig` object in question is created by something not in the MCO codebase? It’s quite possible that we need the MCO change anyway (if the old bug fix was just not working), but to actually fix the reporter’s problem we might need to find and fix that other non-MCO writer.

Forget about my last comment. I was testing on a cluster that build from this patch.
But you are right about the `ContainerRuntimeConfig` object in question is created by something not in the MCO codebase

**Comment by QiWang19** [comment]
The patchContainerRuntimeConfigs can add the `0` that verified through `github.com/clarketm/json.Marshal`  bytes
https://github.com/openshift/machine-config-operator/pull/3836/files#diff-f8f04fec2662cf966bb0a38e978bcf575bcfe7d3630caf7e2a8089d7ec7e6a70R491, what's with the fix https://github.com/openshift/machine-config-operator/pull/2494

**Comment by mtrmac** [comment]
OK, so #2494 was not actually sufficient. Thanks!

/lgtm

**Comment by QiWang19** [comment]
@yuqi-zhang could you approve?

**Review by yuqi-zhang — APPROVED** [review]
logically seems sound. cc @jkyros in case this affects the api move at all. Also doing a

/hold

in case we want to pre-merge verify, but feel free to unhold if we feel that is not necessary

**Comment by yuqi-zhang** [comment]
Sorry for the delay, with https://github.com/openshift/api/pull/1453 and https://github.com/openshift/machine-config-operator/pull/3747, this will first require a change to openshift/api and then re-brought in.

**Pull request closed without merging**
